File: mobile/android/app/src/main/java/com/agrobridge/presentation/screens/transit/TransitTrackingScreen.kt
/**
 * Traceability 2.0 - Real-Time Transit Tracking
 * Android Compose UI
 */

package com.agrobridge.presentation.screens.transit

import android.Manifest
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import com.google.accompanist.permissions.ExperimentalPermissionsApi
import com.google.accompanist.permissions.rememberMultiplePermissionsState
import java.text.SimpleDateFormat
import java.util.*

// MARK: - Models

enum class TransitStatus(
    val displayName: String,
    val color: Long,
    val icon: String
) {
    SCHEDULED("Programado", 0xFF9E9E9E, "schedule"),
    IN_TRANSIT("En tránsito", 0xFF2196F3, "local_shipping"),
    PAUSED("Pausado", 0xFFFF9800, "pause"),
    DELAYED("Retrasado", 0xFFF44336, "warning"),
    COMPLETED("Completado", 0xFF4CAF50, "check_circle"),
    CANCELLED("Cancelado", 0xFF795548, "cancel");

    fun getComposeColor(): Color = Color(color)
}

data class TransitSession(
    val id: String,
    val batchId: String,
    val status: TransitStatus,
    val driverId: String,
    val vehicleId: String?,
    val originName: String,
    val originLat: Double,
    val originLng: Double,
    val destinationName: String,
    val destinationLat: Double,
    val destinationLng: Double,
    val scheduledDeparture: Date?,
    val actualDeparture: Date?,
    val estimatedArrival: Date?,
    val totalDistanceKm: Double?,
    val distanceTraveledKm: Double?,
    val progressPercent: Int = 0
)

// MARK: - Screen

@OptIn(ExperimentalMaterial3Api::class, ExperimentalPermissionsApi::class)
@Composable
fun TransitTrackingScreen(
    viewModel: TransitTrackingViewModel = hiltViewModel(),
    onNavigateBack: () -> Unit = {}
) {
    val uiState by viewModel.uiState.collectAsState()

    val locationPermissions = rememberMultiplePermissionsState(
        permissions = listOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        )
    )

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Tránsito") },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Volver")
                    }
                },
                actions = {
                    if (uiState.activeSession != null) {
                        IconButton(onClick = { /* Open map */ }) {
                            Icon(Icons.Default.Map, contentDescription = "Ver mapa")
                        }
                    }
                }
            )
        }
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            when {
                uiState.isLoading -> {
                    CircularProgressIndicator(modifier = Modifier.align(Alignment.Center))
                }
                uiState.activeSession != null -> {
                    ActiveTransitContent(
                        session = uiState.activeSession!!,
                        isTracking = uiState.isTracking,
                        onStartTransit = {
                            if (locationPermissions.allPermissionsGranted) {
                                viewModel.startTransit()
                            } else {
                                locationPermissions.launchMultiplePermissionRequest()
                            }
                        },
                        onCompleteTransit = { viewModel.completeTransit() }
                    )
                }
                uiState.sessions.isEmpty() -> {
                    EmptyTransitContent(modifier = Modifier.align(Alignment.Center))
                }
                else -> {
                    TransitHistoryList(sessions = uiState.sessions)
                }
            }
        }
    }
}

@Composable
fun ActiveTransitContent(
    session: TransitSession,
    isTracking: Boolean,
    onStartTransit: () -> Unit,
    onCompleteTransit: () -> Unit
) {
    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        // Status Card
        item {
            TransitStatusCard(session = session)
        }

        // Progress
        item {
            if (session.totalDistanceKm != null && session.totalDistanceKm > 0) {
                ProgressCard(
                    progress = session.progressPercent,
                    distanceTraveled = session.distanceTraveledKm ?: 0.0,
                    totalDistance = session.totalDistanceKm
                )
            }
        }

        // Route Info
        item {
            RouteInfoCard(session = session)
        }

        // ETA
        item {
            session.estimatedArrival?.let { eta ->
                ETACard(eta = eta)
            }
        }

        // Actions
        item {
            TransitActionButtons(
                status = session.status,
                isTracking = isTracking,
                onStartTransit = onStartTransit,
                onCompleteTransit = onCompleteTransit
            )
        }
    }
}

@Composable
fun TransitStatusCard(session: TransitSession) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = session.status.getComposeColor().copy(alpha = 0.1f)
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Box(
                modifier = Modifier
                    .size(80.dp)
                    .clip(CircleShape)
                    .background(session.status.getComposeColor().copy(alpha = 0.2f)),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = when (session.status) {
                        TransitStatus.SCHEDULED -> Icons.Default.Schedule
                        TransitStatus.IN_TRANSIT -> Icons.Default.LocalShipping
                        TransitStatus.PAUSED -> Icons.Default.Pause
                        TransitStatus.DELAYED -> Icons.Default.Warning
                        TransitStatus.COMPLETED -> Icons.Default.CheckCircle
                        TransitStatus.CANCELLED -> Icons.Default.Cancel
                    },
                    contentDescription = null,
                    tint = session.status.getComposeColor(),
                    modifier = Modifier.size(40.dp)
                )
            }

            Spacer(modifier = Modifier.height(16.dp))

            Text(
                text = session.status.displayName,
                style = MaterialTheme.typography.headlineSmall,
                fontWeight = FontWeight.Bold
            )
        }
    }
}

@Composable
fun ProgressCard(
    progress: Int,
    distanceTraveled: Double,
    totalDistance: Double
) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.padding(16.dp)) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "$progress% completado",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.Medium
                )
                Text(
                    text = String.format("%.1f / %.1f km", distanceTraveled, totalDistance),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            Spacer(modifier = Modifier.height(8.dp))

            LinearProgressIndicator(
                progress = { progress / 100f },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(8.dp)
                    .clip(RoundedCornerShape(4.dp)),
            )
        }
    }
}

@Composable
fun RouteInfoCard(session: TransitSession) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(modifier = Modifier.padding(16.dp)) {
            // Origin
            Row(verticalAlignment = Alignment.Top) {
                Icon(
                    imageVector = Icons.Default.Circle,
                    contentDescription = null,
                    tint = Color(0xFF4CAF50),
                    modifier = Modifier.size(12.dp)
                )
                Spacer(modifier = Modifier.width(12.dp))
                Column {
                    Text(
                        text = "Origen",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Text(
                        text = session.originName,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Medium
                    )
                }
            }

            // Connector line
            Box(
                modifier = Modifier
                    .padding(start = 5.dp, top = 4.dp, bottom = 4.dp)
                    .width(2.dp)
                    .height(24.dp)
                    .background(MaterialTheme.colorScheme.outline.copy(alpha = 0.3f))
            )

            // Destination
            Row(verticalAlignment = Alignment.Top) {
                Icon(
                    imageVector = Icons.Default.LocationOn,
                    contentDescription = null,
                    tint = Color(0xFFF44336),
                    modifier = Modifier.size(12.dp)
                )
                Spacer(modifier = Modifier.width(12.dp))
                Column {
                    Text(
                        text = "Destino",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Text(
                        text = session.destinationName,
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.Medium
                    )
                }
            }
        }
    }
}

@Composable
fun ETACard(eta: Date) {
    val timeFormat = remember { SimpleDateFormat("HH:mm", Locale.getDefault()) }

    Card(modifier = Modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = Icons.Default.AccessTime,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.primary
            )
            Spacer(modifier = Modifier.width(12.dp))
            Text(
                text = "Llegada estimada:",
                style = MaterialTheme.typography.bodyMedium
            )
            Spacer(modifier = Modifier.weight(1f))
            Text(
                text = timeFormat.format(eta),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.primary
            )
        }
    }
}

@Composable
fun TransitActionButtons(
    status: TransitStatus,
    isTracking: Boolean,
    onStartTransit: () -> Unit,
    onCompleteTransit: () -> Unit
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        when (status) {
            TransitStatus.SCHEDULED -> {
                Button(
                    onClick = onStartTransit,
                    modifier = Modifier.fillMaxWidth(),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = MaterialTheme.colorScheme.primary
                    )
                ) {
                    Icon(Icons.Default.PlayArrow, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Iniciar tránsito")
                }
            }
            TransitStatus.IN_TRANSIT -> {
                Button(
                    onClick = onCompleteTransit,
                    modifier = Modifier.fillMaxWidth(),
                    colors = ButtonDefaults.buttonColors(
                        containerColor = Color(0xFF4CAF50)
                    )
                ) {
                    Icon(Icons.Default.Check, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Completar entrega")
                }

                if (isTracking) {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.Center,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Icon(
                            imageVector = Icons.Default.GpsFixed,
                            contentDescription = null,
                            tint = MaterialTheme.colorScheme.primary,
                            modifier = Modifier.size(16.dp)
                        )
                        Spacer(modifier = Modifier.width(4.dp))
                        Text(
                            text = "Rastreo GPS activo",
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }
            else -> { /* No actions for other states */ }
        }
    }
}

@Composable
fun EmptyTransitContent(modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.padding(32.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Icon(
            imageVector = Icons.Default.LocalShipping,
            contentDescription = null,
            modifier = Modifier.size(80.dp),
            tint = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.5f)
        )
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Sin sesiones de tránsito",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.SemiBold
        )
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "Crea una sesión desde la app de administración",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
fun TransitHistoryList(sessions: List<TransitSession>) {
    LazyColumn(
        contentPadding = PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(sessions, key = { it.id }) { session ->
            TransitSessionRow(session = session)
        }
    }
}

@Composable
fun TransitSessionRow(session: TransitSession) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Icon(
                imageVector = when (session.status) {
                    TransitStatus.COMPLETED -> Icons.Default.CheckCircle
                    TransitStatus.CANCELLED -> Icons.Default.Cancel
                    else -> Icons.Default.LocalShipping
                },
                contentDescription = null,
                tint = session.status.getComposeColor(),
                modifier = Modifier.size(32.dp)
            )

            Spacer(modifier = Modifier.width(12.dp))

            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = "${session.originName} → ${session.destinationName}",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.Medium,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
                Text(
                    text = session.status.displayName,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            session.totalDistanceKm?.let { distance ->
                Text(
                    text = String.format("%.0f km", distance),
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }
    }
}

File: apps/api/docs/mobile/TransitTrackingScreen.swift
/**
 * Traceability 2.0 - Real-Time Transit Tracking
 * iOS SwiftUI Implementation
 *
 * Features:
 * - View transit sessions for a batch
 * - Start/complete transit
 * - Real-time GPS location updates
 * - Map view with route and current position
 * - Route deviation alerts
 */

import SwiftUI
import MapKit
import CoreLocation

// MARK: - Models

enum TransitStatus: String, Codable {
    case SCHEDULED, IN_TRANSIT, PAUSED, DELAYED, COMPLETED, CANCELLED

    var displayName: String {
        switch self {
        case .SCHEDULED: return "Programado"
        case .IN_TRANSIT: return "En tránsito"
        case .PAUSED: return "Pausado"
        case .DELAYED: return "Retrasado"
        case .COMPLETED: return "Completado"
        case .CANCELLED: return "Cancelado"
        }
    }

    var color: Color {
        switch self {
        case .SCHEDULED: return .gray
        case .IN_TRANSIT: return .blue
        case .PAUSED: return .orange
        case .DELAYED: return .red
        case .COMPLETED: return .green
        case .CANCELLED: return .brown
        }
    }

    var icon: String {
        switch self {
        case .SCHEDULED: return "clock"
        case .IN_TRANSIT: return "truck.box"
        case .PAUSED: return "pause.circle"
        case .DELAYED: return "exclamationmark.triangle"
        case .COMPLETED: return "checkmark.circle"
        case .CANCELLED: return "xmark.circle"
        }
    }
}

struct TransitSession: Identifiable, Codable {
    let id: String
    let batchId: String
    let status: TransitStatus
    let driverId: String
    let vehicleId: String?
    let originName: String
    let originLat: Double
    let originLng: Double
    let destinationName: String
    let destinationLat: Double
    let destinationLng: Double
    let scheduledDeparture: Date?
    let actualDeparture: Date?
    let scheduledArrival: Date?
    let actualArrival: Date?
    let estimatedArrival: Date?
    let totalDistanceKm: Double?
    let distanceTraveledKm: Double?
    let progressPercent: Int?
    let lastLocation: TransitLocation?
}

struct TransitLocation: Identifiable, Codable {
    let id: String
    let sessionId: String
    let latitude: Double
    let longitude: Double
    let speed: Double?
    let heading: Double?
    let isOffRoute: Bool
    let deviationKm: Double?
    let timestamp: Date
}

// MARK: - ViewModel

@MainActor
class TransitTrackingViewModel: NSObject, ObservableObject, CLLocationManagerDelegate {
    @Published var sessions: [TransitSession] = []
    @Published var activeSession: TransitSession?
    @Published var locations: [TransitLocation] = []
    @Published var isLoading = false
    @Published var error: String?
    @Published var isTracking = false

    private let locationManager = CLLocationManager()
    private let batchId: String
    private let authToken: String
    private let baseURL: String
    private var locationTimer: Timer?

    init(batchId: String, authToken: String) {
        self.batchId = batchId
        self.authToken = authToken
        self.baseURL = ProcessInfo.processInfo.environment["API_BASE_URL"] ?? "https://api.agrobridge.com"
        super.init()

        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
    }

    func loadSessions() async {
        isLoading = true
        error = nil

        do {
            guard let url = URL(string: "\(baseURL)/api/v1/batches/\(batchId)/transit") else {
                throw URLError(.badURL)
            }

            var request = URLRequest(url: url)
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

            let (data, _) = try await URLSession.shared.data(for: request)

            struct Response: Codable {
                let success: Bool
                let data: DataWrapper
            }
            struct DataWrapper: Codable {
                let sessions: [TransitSession]
            }

            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            let result = try decoder.decode(Response.self, from: data)
            sessions = result.data.sessions

            // Find active session
            activeSession = sessions.first {
                [.SCHEDULED, .IN_TRANSIT, .PAUSED, .DELAYED].contains($0.status)
            }
        } catch {
            self.error = "Error loading sessions: \(error.localizedDescription)"
        }

        isLoading = false
    }

    func startTransit() async {
        guard let session = activeSession else { return }

        do {
            guard let url = URL(string: "\(baseURL)/api/v1/transit/\(session.id)/start") else { return }

            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

            let (data, _) = try await URLSession.shared.data(for: request)

            struct Response: Codable {
                let data: TransitSession
            }
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            let result = try decoder.decode(Response.self, from: data)
            activeSession = result.data

            // Start location tracking
            startLocationTracking()
        } catch {
            self.error = "Error starting transit: \(error.localizedDescription)"
        }
    }

    func completeTransit() async {
        guard let session = activeSession else { return }

        do {
            guard let url = URL(string: "\(baseURL)/api/v1/transit/\(session.id)/complete") else { return }

            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")

            let (data, _) = try await URLSession.shared.data(for: request)

            struct Response: Codable {
                let data: TransitSession
            }
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            let result = try decoder.decode(Response.self, from: data)
            activeSession = result.data

            stopLocationTracking()
        } catch {
            self.error = "Error completing transit: \(error.localizedDescription)"
        }
    }

    func startLocationTracking() {
        locationManager.requestWhenInUseAuthorization()
        locationManager.startUpdatingLocation()
        isTracking = true

        // Send location updates every 30 seconds
        locationTimer = Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { [weak self] _ in
            self?.sendLocationUpdate()
        }
    }

    func stopLocationTracking() {
        locationManager.stopUpdatingLocation()
        locationTimer?.invalidate()
        locationTimer = nil
        isTracking = false
    }

    private func sendLocationUpdate() {
        guard let session = activeSession,
              let location = locationManager.location else { return }

        Task {
            do {
                guard let url = URL(string: "\(baseURL)/api/v1/transit/\(session.id)/location") else { return }

                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")

                let body: [String: Any] = [
                    "latitude": location.coordinate.latitude,
                    "longitude": location.coordinate.longitude,
                    "altitude": location.altitude,
                    "accuracy": location.horizontalAccuracy,
                    "speed": max(0, location.speed * 3.6), // Convert m/s to km/h
                    "heading": location.course
                ]
                request.httpBody = try JSONSerialization.data(withJSONObject: body)

                let (_, _) = try await URLSession.shared.data(for: request)
            } catch {
                print("Error sending location: \(error)")
            }
        }
    }

    // CLLocationManagerDelegate
    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        // Location updates handled by timer
    }
}

// MARK: - Views

struct TransitTrackingScreen: View {
    @StateObject var viewModel: TransitTrackingViewModel
    @State private var showMap = false

    var body: some View {
        NavigationView {
            Group {
                if viewModel.isLoading && viewModel.sessions.isEmpty {
                    ProgressView("Cargando sesiones...")
                } else if let session = viewModel.activeSession {
                    ActiveTransitView(session: session, viewModel: viewModel)
                } else if viewModel.sessions.isEmpty {
                    EmptyTransitView()
                } else {
                    TransitHistoryList(sessions: viewModel.sessions)
                }
            }
            .navigationTitle("Tránsito")
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    if viewModel.activeSession != nil {
                        Button(action: { showMap = true }) {
                            Image(systemName: "map")
                        }
                    }
                }
            }
            .sheet(isPresented: $showMap) {
                if let session = viewModel.activeSession {
                    TransitMapView(session: session)
                }
            }
        }
        .task {
            await viewModel.loadSessions()
        }
    }
}

struct ActiveTransitView: View {
    let session: TransitSession
    @ObservedObject var viewModel: TransitTrackingViewModel

    var body: some View {
        ScrollView {
            VStack(spacing: 24) {
                // Status Card
                TransitStatusCard(session: session)

                // Progress
                if let progress = session.progressPercent {
                    ProgressView(value: Double(progress) / 100) {
                        HStack {
                            Text("\(progress)% completado")
                            Spacer()
                            if let remaining = session.totalDistanceKm.map({ $0 - (session.distanceTraveledKm ?? 0) }) {
                                Text(String(format: "%.1f km restantes", remaining))
                                    .foregroundColor(.secondary)
                            }
                        }
                    }
                    .padding()
                }

                // Route Info
                RouteInfoCard(session: session)

                // ETA
                if let eta = session.estimatedArrival {
                    HStack {
                        Image(systemName: "clock")
                            .foregroundColor(.blue)
                        Text("Llegada estimada:")
                        Spacer()
                        Text(eta, style: .time)
                            .fontWeight(.medium)
                    }
                    .padding()
                    .background(Color(.systemGray6))
                    .cornerRadius(12)
                    .padding(.horizontal)
                }

                // Actions
                TransitActionButtons(session: session, viewModel: viewModel)
            }
            .padding(.vertical)
        }
    }
}

struct TransitStatusCard: View {
    let session: TransitSession

    var body: some View {
        VStack(spacing: 16) {
            ZStack {
                Circle()
                    .fill(session.status.color.opacity(0.2))
                    .frame(width: 80, height: 80)

                Image(systemName: session.status.icon)
                    .font(.system(size: 36))
                    .foregroundColor(session.status.color)
            }

            Text(session.status.displayName)
                .font(.title2)
                .fontWeight(.bold)
        }
        .padding()
    }
}

struct RouteInfoCard: View {
    let session: TransitSession

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack(alignment: .top) {
                Image(systemName: "circle.fill")
                    .foregroundColor(.green)
                    .font(.caption)
                VStack(alignment: .leading) {
                    Text("Origen")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    Text(session.originName)
                        .fontWeight(.medium)
                }
                Spacer()
            }

            Rectangle()
                .fill(Color.gray.opacity(0.3))
                .frame(width: 2, height: 30)
                .padding(.leading, 4)

            HStack(alignment: .top) {
                Image(systemName: "mappin.circle.fill")
                    .foregroundColor(.red)
                    .font(.caption)
                VStack(alignment: .leading) {
                    Text("Destino")
                        .font(.caption)
                        .foregroundColor(.secondary)
                    Text(session.destinationName)
                        .fontWeight(.medium)
                }
                Spacer()
            }
        }
        .padding()
        .background(Color(.systemGray6))
        .cornerRadius(12)
        .padding(.horizontal)
    }
}

struct TransitActionButtons: View {
    let session: TransitSession
    @ObservedObject var viewModel: TransitTrackingViewModel

    var body: some View {
        VStack(spacing: 12) {
            if session.status == .SCHEDULED {
                Button(action: {
                    Task { await viewModel.startTransit() }
                }) {
                    Label("Iniciar tránsito", systemImage: "play.fill")
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.blue)
                        .foregroundColor(.white)
                        .cornerRadius(12)
                }
            }

            if session.status == .IN_TRANSIT {
                Button(action: {
                    Task { await viewModel.completeTransit() }
                }) {
                    Label("Completar entrega", systemImage: "checkmark")
                        .frame(maxWidth: .infinity)
                        .padding()
                        .background(Color.green)
                        .foregroundColor(.white)
                        .cornerRadius(12)
                }

                if viewModel.isTracking {
                    HStack {
                        Image(systemName: "location.fill")
                            .foregroundColor(.blue)
                        Text("Rastreo GPS activo")
                            .foregroundColor(.secondary)
                    }
                    .font(.caption)
                }
            }
        }
        .padding(.horizontal)
    }
}

struct TransitMapView: View {
    let session: TransitSession
    @State private var region: MKCoordinateRegion

    init(session: TransitSession) {
        self.session = session
        let center = CLLocationCoordinate2D(
            latitude: (session.originLat + session.destinationLat) / 2,
            longitude: (session.originLng + session.destinationLng) / 2
        )
        _region = State(initialValue: MKCoordinateRegion(
            center: center,
            span: MKCoordinateSpan(latitudeDelta: 0.5, longitudeDelta: 0.5)
        ))
    }

    var body: some View {
        Map(coordinateRegion: $region, annotationItems: annotations) { item in
            MapAnnotation(coordinate: item.coordinate) {
                Image(systemName: item.icon)
                    .foregroundColor(item.color)
                    .padding(8)
                    .background(Circle().fill(.white))
                    .shadow(radius: 2)
            }
        }
        .ignoresSafeArea()
    }

    var annotations: [MapPoint] {
        var points: [MapPoint] = [
            MapPoint(
                id: "origin",
                coordinate: CLLocationCoordinate2D(latitude: session.originLat, longitude: session.originLng),
                icon: "circle.fill",
                color: .green
            ),
            MapPoint(
                id: "destination",
                coordinate: CLLocationCoordinate2D(latitude: session.destinationLat, longitude: session.destinationLng),
                icon: "mappin.circle.fill",
                color: .red
            )
        ]

        if let location = session.lastLocation {
            points.append(MapPoint(
                id: "current",
                coordinate: CLLocationCoordinate2D(latitude: location.latitude, longitude: location.longitude),
                icon: "truck.box.fill",
                color: .blue
            ))
        }

        return points
    }
}

struct MapPoint: Identifiable {
    let id: String
    let coordinate: CLLocationCoordinate2D
    let icon: String
    let color: Color
}

struct EmptyTransitView: View {
    var body: some View {
        VStack(spacing: 20) {
            Image(systemName: "truck.box")
                .font(.system(size: 60))
                .foregroundColor(.gray)
            Text("Sin sesiones de tránsito")
                .font(.title2)
            Text("Crea una sesión desde la app de administración")
                .foregroundColor(.secondary)
        }
    }
}

struct TransitHistoryList: View {
    let sessions: [TransitSession]

    var body: some View {
        List(sessions) { session in
            TransitSessionRow(session: session)
        }
        .listStyle(.insetGrouped)
    }
}

struct TransitSessionRow: View {
    let session: TransitSession

    var body: some View {
        HStack {
            Image(systemName: session.status.icon)
                .foregroundColor(session.status.color)
                .frame(width: 40)

            VStack(alignment: .leading) {
                Text("\(session.originName) → \(session.destinationName)")
                    .fontWeight(.medium)
                    .lineLimit(1)
                Text(session.status.displayName)
                    .font(.caption)
                    .foregroundColor(.secondary)
            }

            Spacer()

            if let distance = session.totalDistanceKm {
                Text(String(format: "%.0f km", distance))
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
        }
        .padding(.vertical, 4)
    }
}
